### DISCUSSION TITLE
mmcqd not started by flashed kernel partition

### DISCUSSION THREAD

**Discussion by archandanime** [gtxaspec/wz_mini_hacks#150] [Development]
I flashed kernel by a bootable kernel, the same kernel as factory_t31_ZMC6tiIDQN that I booted smoothly before flashing. It's it's small enough to fit the partition. However, the factory_t31_ZMC6tiIDQN can detect the SD card but flashed kernel can't.
When I insert the SD card:
flashed kernel:
```
[    0.823563] jzmmc_v1.2 jzmmc_v1.2.0: enter jzmmc detect change cd-pin 0
[    1.023553] jzmmc_v1.2 jzmmc_v1.2.0: enter jzmmc detect change cd-pin 0
/ # 
```
factory_t31_ZMC6tiIDQN:
```
[    0.893562] mmc0: new high speed SDHC card at address 5048
[    0.899442] mmcblk0: mmc0:5048 SD16G 14.4 GiB 
[    0.905390]  mmcblk0: p1
```
After looking around, I notice the difference: there is mmcqd process run when factory_t31_ZMC6tiIDQN is booted, but there's no such process by flashed kernel
factory_t31_ZMC6tiIDQN:
```
/ # ps
  PID USER       VSZ STAT COMMAND
    1 0          868 S    /bin/sh
    2 0            0 SW   [kthreadd]
    3 0            0 SW   [ksoftirqd/0]
    4 0            0 SW   [kworker/0:0]
    5 0            0 SW<  [kworker/0:0H]
    6 0            0 SW   [kworker/u2:0]
    7 0            0 SW   [rcu_preempt]
    8 0            0 SW   [rcu_bh]
    9 0            0 SW   [rcu_sched]
   10 0            0 SW   [watchdog/0]
   11 0            0 SW<  [khelper]
   12 0            0 SW   [kdevtmpfs]
   13 0            0 SW<  [writeback]
   14 0            0 SW<  [bioset]
   15 0            0 SW<  [kblockd]
   16 0            0 SW   [kworker/0:1]
   17 0            0 SW<  [cfg80211]
   18 0            0 SW   [kswapd0]
   19 0            0 SW   [fsnotify_mark]
   20 0            0 SW<  [crypto]
   34 0            0 SW   [kworker/u2:1]
   35 0            0 SW   [kworker/u2:2]
   36 0            0 SW<  [deferwq]
   44 0            0 SW   [mmcqd/0]
   46 0            0 SW<  [kworker/0:1H]
   47 0            0 SW   [kworker/0:2]
```
flashed kernel:
```
/ # ps
  PID USER       VSZ STAT COMMAND
    1 0          868 S    /bin/sh
    2 0            0 SW   [kthreadd]
    3 0            0 SW   [ksoftirqd/0]
    4 0            0 SW   [kworker/0:0]
    5 0            0 SW<  [kworker/0:0H]
    6 0            0 SW   [kworker/u2:0]
    7 0            0 SW   [rcu_preempt]
    8 0            0 SW   [rcu_bh]
    9 0            0 SW   [rcu_sched]
   10 0            0 SW   [watchdog/0]
   11 0            0 SW<  [khelper]
   12 0            0 SW   [kdevtmpfs]
   13 0            0 SW<  [writeback]
   14 0            0 SW<  [bioset]
   15 0            0 SW<  [kblockd]
   16 0            0 SW   [kworker/0:1]
   17 0            0 SW<  [cfg80211]
   18 0            0 SW   [kswapd0]
   19 0            0 SW   [fsnotify_mark]
   20 0            0 SW<  [crypto]
   34 0            0 SW   [kworker/u2:1]
   35 0            0 SW<  [deferwq]
   44 0            0 SW<  [kworker/0:1H]
   45 0            0 SW   [kworker/0:2]
   46 0          864 R    ps
```
they are same kernel image, are they supposed to behave the same way?
could you give me your opinion about this situation please?
thanks

**Comment by gtxaspec** [comment]
what options did you change in the kernel config? the mmc driver may not be compiled into the kernel...
how are you flashing the kernel?

**Comment by archandanime** [comment]
I flashed /dev/mtdblock1 with dd. I got rid of "USB support" and "Real time clock" the reduce kernel size

**Comment by archandanime** [comment]
"MMC/SD/SDIO card support" and everything else are untouched

**Comment by gtxaspec** [comment]
dd will not work properly.  you need to use `flashcp -v <image> /dev/mtd1`

**Comment by archandanime** [comment]
I ran the command
```
/sdcard # busybox flashcp -v factory_t31_ZMC6tiIDQN /dev/mtd1
Erasing block: 62/62 (100%) 
Writing kb: 1982/1982 (100%) 
Verifying kb: 1982/1982 (100%) 
/sdcard # 
```
removed sd card and rebooted, it still can't detect sd card, no mmcqd running, no /dev/mmc*

**Comment by gtxaspec** [comment]
does `/proc/jz/mmc0` exist on the flashed kernel?

do you see this in early dmesg?
```
[    0.676518] jzmmc_v1.2 jzmmc_v1.2.0: vmmc regulator missing
[    0.682509] jzmmc_v1.2 jzmmc_v1.2.0: register success!
[    0.687857] jzmmc_v1.2 jzmmc_v1.2.0: card inserted, state=0
[    0.693707] jzmmc_v1.2 jzmmc_v1.2.1: vmmc regulator missing
[    0.699669] jzmmc_v1.2 jzmmc_v1.2.1: register success!
```

**Comment by archandanime** [comment]
Edited: `/proc/jz/mmc0` does exist in flashed kernel!
```
/ # ls /proc/jz/
clock  ddr    debug  gpio   mmc0   reset
/ # ls /dev/mmc*
ls: /dev/mmc*: No such file or directory
/ # 
```
/proc/jz/mmc0 dir is empty
yes. there were such messages
```
[    0.615480] jzmmc_v1.2 jzmmc_v1.2.0: vmmc regulator missing
[    0.621465] jzmmc_v1.2 jzmmc_v1.2.0: register success!
[    0.626810] jzmmc_v1.2 jzmmc_v1.2.0: card inserted, state=0
[    0.632667] jzmmc_v1.2 jzmmc_v1.2.1: vmmc regulator missing
[    0.638596] jzmmc_v1.2 jzmmc_v1.2.1: register success!
```

**Comment by gtxaspec** [comment]
thats really weird.  maybe you need to enable the gpio.

i think its 50 or 48

`echo 50 > /sys/class/gpio/export`
`echo in > /sys/class/gpio/gpio50/direction`

if 50 doesnt work try 58

**Comment by archandanime** [comment]
```
/ # export 50 > /sys/class/gpio/export
/bin/sh: export: 50: bad variable name
/ #  echo in > /sys/class/gpio/gpio50/direction
/bin/sh: can't create /sys/class/gpio/gpio50/direction: nonexistent directory
```

**Reply by gtxaspec** [reply]
oops, its:


`echo 50 > /sys/class/gpio/export`
`echo in > /sys/class/gpio/gpio50/direction`

**Comment by archandanime** [comment]
```
/ # echo 50 > /sys/class/gpio/export
/ # [    4.943557] jzmmc_v1.2 jzmmc_v1.2.0: enter jzmmc detect change cd-pin 1

/ #  echo in > /sys/class/gpio/gpio50/direction
/ # [   12.263557] jzmmc_v1.2 jzmmc_v1.2.0: enter jzmmc detect change cd-pin 0
[   12.463547] jzmmc_v1.2 jzmmc_v1.2.0: enter jzmmc detect change cd-pin 0
```

**Reply by gtxaspec** [reply]
try 48 and 59

**Comment by archandanime** [comment]
```
/ # echo 48 > /sys/class/gpio/export
/ #  echo in > /sys/class/gpio/gpio48/direction
/ # echo 59 > /sys/class/gpio/export
sh: write error: Device or resource busy
/ #  echo in > /sys/class/gpio/gpio59/direction
/bin/sh: can't create /sys/class/gpio/gpio59/direction: nonexistent directory
/ # 
```

**Comment by gtxaspec** [comment]
hmm...this is interesting.  I haven't flashed kernels in a long long time.  Can you try flashing the stock wz_mini kernel... does that work?

**Comment by archandanime** [comment]
stock wz_mini kernel is quite big, I'm not sure if it works, but I can try. I'm backing up partitions first

**Comment by gtxaspec** [comment]
oh yeah i forgot 1984k limit. how big is it?

**Comment by gtxaspec** [comment]
disable:
`CONFIG_KALLSYMS` in the kernel config.  this will reduce it by a bunch too.

**Reply by archandanime** [reply]
should I disabled CONFIG_KALLSYMS_ALL too?

**Reply by gtxaspec** [reply]
yeah disable both

**Comment by archandanime** [comment]
```
/sdcard # dd if=/dev/mtdblock1 of=mtdblock1
3968+0 records in
3968+0 records out
/sdcard # ls -li factory_t31_ZMC6tiIDQN_wzmini mtdblock1
    110 -rwxr-xr-x    1 0        0          2105727 Jun 28  2022 factory_t31_ZMC6tiIDQN_wzmini
    119 -rwxr-xr-x    1 0        0          2031616 Jan  1 00:01 mtdblock1
```

**Comment by archandanime** [comment]
I also tried to re-compile busybox but it's never smaller than mnakada's `initramfs_skeleton/bin/busybox`(758782 bytes)

**Reply by gtxaspec** [reply]
the busybox i compiled is smaller than mnakada's...i used musl and its the smallest i could get it lol

**Reply by gtxaspec** [reply]
`559468 Jun 30 10:27 busybox`

**Reply by archandanime** [reply]
how small is it, could you share the config please?

**Reply by gtxaspec** [reply]
[busybox.gz](https://github.com/gtxaspec/wz_mini_hacks/files/9024887/busybox.gz)

**Reply by archandanime** [reply]
thank you very much!

**Comment by archandanime** [comment]
```
/sdcard # busybox flashcp -v factory_t31_ZMC6tiIDQN_wzmini /dev/mtd1
flashcp: factory_t31_ZMC6tiIDQN_wzmini bigger than /dev/mtd1
```

**Reply by gtxaspec** [reply]
how big is the kernel?

**Comment by archandanime** [comment]
```
/sdcard # dd if=/dev/mtdblock1 of=mtdblock1
3968+0 records in
3968+0 records out
/sdcard # ls -li factory_t31_ZMC6tiIDQN_wzmini mtdblock1
    110 -rwxr-xr-x    1 0        0          2105727 Jun 28  2022 factory_t31_ZMC6tiIDQN_wzmini
    119 -rwxr-xr-x    1 0        0          2031616 Jan  1 00:01 mtdblock1
```
71kb bigger

**Reply by gtxaspec** [reply]
remove netfilter, and remove ipv6 and see if that gets it under the size.  is that with kallsyms out? its the same size you shared before.

**Comment by archandanime** [comment]
it's now 1964660,

**Comment by archandanime** [comment]
I flashed it, I don't see anything new

**Reply by gtxaspec** [reply]
rtc and usb enabled? still no mmc?

**Reply by archandanime** [reply]
I only replaced kernel.config and init busybox, no change to kernel.config at all, init is custom

**Comment by gtxaspec** [comment]
[kernel.bin.gz](https://github.com/gtxaspec/wz_mini_hacks/files/9024970/kernel.bin.gz)
try this

**Comment by archandanime** [comment]
it boots to stock, I can see the view on Wyze app.
with initramfs debug enabled. there IS /dev/mmcblk0p1

**Comment by gtxaspec** [comment]
you flashed the kernel?

**Comment by archandanime** [comment]
I flashed your kernel.bin, deleted `factory_t31_ZMC6tiIDQN` on SD card and inserted it

**Comment by gtxaspec** [comment]
it must be something with your kernel config then that could be wonky.  Try using the stock one for wz_mini and try again

**Comment by archandanime** [comment]
I'm not very familiar with docker, lemme delete the container and image and start over. it will take a while

**Reply by gtxaspec** [reply]
use lxc.  do you have ubuntu?

**Reply by archandanime** [reply]
I use arch, having been using for years

**Reply by gtxaspec** [reply]
i use lxc, then use an ubuntu based container for speed, and use that as my development environment.  faster than docker.

**Reply by archandanime** [reply]
thank you for the suggestion! I'll definitely do it, I have a spare partition for fedora but never use it

**Reply by gtxaspec** [reply]
yah, then you can do `lxc-create -t download -n t31-devel`
select ubuntu, jammy, amd64 or whatever your host arch is

**Reply by archandanime** [reply]
does it work with mnakada's repo too? I saw some lines with docker on his Makefile. Is replacing these lines all I need?

**Reply by gtxaspec** [reply]
you can run docker inside the lxc container, thats what i do, just for his atomcam_tools project

**Reply by archandanime** [reply]
that sounds awesome, thank you! I'm deleting docker :)). One more question: it works well with Arch like on Ubuntu, right? my Arch system is on SSD along with Windows. Using another OS on HDD is kinda slow

**Reply by gtxaspec** [reply]
you should be able to install it in arch: https://wiki.archlinux.org/title/Linux_Containers

**Reply by gtxaspec** [reply]
set up a privileged container to run docker

**Reply by archandanime** [reply]
thank you!

**Reply by archandanime** [reply]
another question please, can I delete mipsel-linux-gnu- cross-compiling tools if I already have lxc?

**Reply by gtxaspec** [reply]
you will need to download them in the container.  all your tools and everything go in the container
 https://github.com/Dafang-Hacks/mips-gcc472-glibc216-64bit/archive/master.zip

**Reply by archandanime** [reply]
lovely! thank you!

**Comment by archandanime** [comment]
hi
I installed lxc and was so excited to see its potential, it took me a day to config it
I made both arch and ubuntu containers(both unprivileged) , however when I tried to run docker inside both containers it failed:
```
[root@arch atomcam_tools]$  make
latest: Pulling from atomtools/atomtools
Digest: sha256:ffbcf491f8536ac50a3be2fd01f56ba9b21ecb4eccac8dd85c5b61c36d05b2d6
Status: Image is up to date for atomtools/atomtools:latest
docker.io/atomtools/atomtools:latest
[+] Running 0/1
 ⠹ Container atomcam_tools-builder-1  Starting                             2.3s
Error response from daemon: failed to create shim task: OCI runtime create failed: runc create failed: unable to start container process: can't get final child's PID from pipe: EOF: unknown
make: *** [Makefile:6: build] Error 1
```
`sysctl -a`
```
user.max_cgroup_namespaces = 31077
user.max_pid_namespaces = 31077
```
`cat /proc/cmdline`
```
... quiet loglevel=3 rd.systemd.show_status=auto systemd.unified_cgroup_hierarchy=1 namespace.unpriv_enable=1 rd.udev.log_level=3 initrd=\initramfs-linux.img
```
`lxc-info arch`
```
Name:           arch
State:          RUNNING
PID:            1564
IP:             10.0.3.177
IP:             172.17.0.1
IP:             172.18.0.1
Link:           veth1000_gnPV
 TX bytes:      10.73 KiB
 RX bytes:      27.58 KiB
 Total bytes:   38.31 KiB
```
`lxc-checkconfig`
```
LXC version 5.0.0
--- Namespaces ---
Namespaces: enabled
Utsname namespace: enabled
Ipc namespace: enabled
Pid namespace: enabled
User namespace: enabled
Network namespace: enabled

--- Control groups ---
Cgroups: enabled
Cgroup namespace: enabled

Cgroup v1 mount points: 


Cgroup v2 mount points: 
/sys/fs/cgroup

Cgroup v1 systemd controller: missing
Cgroup v1 freezer controller: missing
Cgroup ns_cgroup: required
Cgroup device: enabled
Cgroup sched: enabled
Cgroup cpu account: enabled
Cgroup memory controller: enabled
Cgroup cpuset: enabled

--- Misc ---
Veth pair device: enabled, loaded
Macvlan: enabled, not loaded
Vlan: enabled, not loaded
Bridges: enabled, loaded
Advanced netfilter: enabled, loaded
CONFIG_IP_NF_TARGET_MASQUERADE: enabled, not loaded
CONFIG_IP6_NF_TARGET_MASQUERADE: enabled, not loaded
CONFIG_NETFILTER_XT_TARGET_CHECKSUM: enabled, not loaded
CONFIG_NETFILTER_XT_MATCH_COMMENT: enabled, not loaded
FUSE (for use with lxcfs): enabled, loaded

--- Checkpoint/Restore ---
checkpoint restore: enabled
CONFIG_FHANDLE: enabled
CONFIG_EVENTFD: enabled
CONFIG_EPOLL: enabled
CONFIG_UNIX_DIAG: enabled
CONFIG_INET_DIAG: enabled
CONFIG_PACKET_DIAG: enabled
CONFIG_NETLINK_DIAG: enabled
File capabilities: 

Note : Before booting a new kernel, you can check its configuration
usage : CONFIG=/path/to/config /usr/bin/lxc-checkconfig
```
restarting docker/containerd didn't help. May I ask for your advice please?

**Comment by gtxaspec** [comment]
The config file for this priveleged container in /var/lib/lxc/container_name/config, Add the following lines:
```
# For docker
lxc.apparmor.profile = unconfined
```

**Comment by archandanime** [comment]
I added `lxc.apparmor.profile = unconfined` to `.local/share/lxc/arch/config` and `.local/share/lxc/ubuntu/config`, it's didn't solve
I don't have apparmor installed on host machine, do I need it to make docker in the container work?

**Reply by gtxaspec** [reply]
if you have appamor installed you need that line on the host, and docker will only work in a privileged container btw

**Reply by archandanime** [reply]
I see, thank you for pointing it out
